broadinstitute/single_cell_portal_core
# Remove extended OAuth scopes, use Terra pet service accounts (SCP-2318, SCP-3204, SCP-3205, SCP-3250, SCP-3259)

Previously, when users authenticate into SCP, we asked for two "extended" OAuth scopes: devstorage.read_only, and cloud-billing.readonly.  The former was to allow streaming of files directly to the browser in private studies (for IGV, Ideogram), and the latter to retrieve the list of a user's GCP billing accounts when they are managing Terra billing projects through the "My Billing Projects" section of SCP.

Terra has since stopped requesting extended scopes from users, as some users are unwilling or unable to grant them, and this effectively prevents them from using Terra.  Similar issues have arisen lately with SCP users who object to (or cannot grant) these scope requests.

This update removes both extended scopes from the normal SCP login process, and replaces them with the following functionality:
- **devstorage.read_only**: SCP now leverages Terra pet service accounts, which are Terra-managed service accounts that can be used to grant access tokens w/ extended scopes.  Now, when a user attempts to stream a file directly to the browser in a private study they have access to, SCP issues a token using this pet service account, rather than directly using the user's access token, as it no longer has the requisite scope. For users who have completed their Terra profile, this flow will work.  Users who have not registered with Terra must first complete their Terra registration, as this request will result in a `401` from SAM.
- **cloud-billing.readonly**: If a user clicks the link for "My Billing Projects", they are redirected to a page informing them of the extended scope requirement, and asks them to re-authenticate and grant this scope.  They then have the choice of agreeing or declining at this point.  Users who are unable/unwilling to grant this scope cannot manage billing projects through SCP, but they are no longer blocked from signing into SCP altogether.

This update also displays warnings for users who have not completed their Terra registration when attempting to load IGV or Ideogram in private studies.  Furthermore, as a part of SCP-3259, users can no longer update their Terra profile through SCP until they complete their Terra registration and accept their terms of service.

To test:
1. Pull this branch and sign out of any active sessions
2. Sign in as normal with an account that has registered for Terra
3. Note that no extra consent is required
4. Click the "My Billing Projects" link from the top-righthand nav
5. Note the landing page explaining the extra scope requirements
6. Click 'Sign in and Grant Access'
7. Note the cloud billing scope request
8. Consent to the request, note that the billing projects page loads

If you have any studies with BAM files or Ideogram annotations, to test this functionality:
1. Follow steps 1-3 from above
2. Load the study w/ IGV or Ideogram files
3. Note that the visualization loads
4. Sign out
5. Sign in with an account that has not registered for Terra
6. Load the previous study
7. Note the warning dialog prompting to complete your Terra registration

This PR satisfies SCP-2318, SCP-3204, SCP-3205, SCP-3250, and SCP-3259.
